[draft] Intro to Reactivity (#323)
solidjs/solid-docs-next: content/essentials/intro-to-reactivity.mdx

### Signals

Signals serve as foundational elements in reactive systems, playing a pivotal role in data management and system responsiveness.

A signal can be visualized as a container for data. 
It not only holds the data, but also provides mechanisms for accessing and modifying its value, a **getter **and a **setter**. 
This dual capability makes signals both readable and mutable, making them integral to the dynamic nature of reactive systems.

#### Getter

The getter is the mechanism responsible for accessing the current value of the signal. 
When various components or parts of an application need to "read" or retrieve the data stored in a signal, they use the getter. 
This provides data that is consistent and reliably accessible throughout the system.

#### Setter

On the other hand, the setter is the tool that permits the changing of the signal's value. 
Any change to the value within a signal, made possible by the setter, can instigate reactive updates in other connected parts of the system. 
This ensures that the system remains responsive and up-to-date with the latest data changes.

### Effects

> Effects can be thought of as automated responder. 

Reviewer: Maybe something like
> Effects and other listeners (such as memos) can be thought of as automated responders.

because effects are not special in this regard.

The altered signal proactively notifies all its subscribed effects about the modification. 
Upon receiving this notification, these effects then update or activate as necessary to keep the system's state current and synchronized with the latest data changes.

## Synchronous vs. asynchronous

Reactivity in systems can manifest primarily in two ways: synchronous and asynchronous. 
The choice between these often depend on the requirements of the system and the nature of the tasks involved.

### Synchronous reactivity
Synchronous reactivity operates in a direct and linear fashion. 
When a signal changes, its corresponding effects are immediately updated in a sequential manner. 
This form of reactivity provides a direct correlation between a signal's change and the system's response, guaranteeing that the system remains consistently aligned with the most recent data.

### Asynchronous reactivity
Asynchronous reactivity introduces a layer of flexibility and adaptability. 
Unlike it synchronous counterpart, the updates to the effects are not instantaneous. 
Instead, they may be delayed or might wait for the completion of an external event or task. 

> This approach is particularly useful in scenarios where immediate updates are not feasible or when the system needs to accommodate external processes before reflecting changes, such as waiting for an API response.

Reviewer: I asked in the discord:
> Do we have an example of asynchronous reactivity in solid? At first I thought it was referring to batch, but now I think it is referring to suspense. However, batch is more about making sure all state is updated before allowing effects to run, and suspense isn't about delaying computations at all.

In any case, assuming this is referring to batching, I would write it like this:
> This approach is useful in scenarios where an effect depends on multiple signals, and updating one signal before the other would result a data inconsistency. In such a scenario, `batch` can be used to delay triggering effects, so the effect runs after both signals have been updated.

Author: do you think its beneficial to talk about both suspense and batch in this case?

In the context of reactivity, memos become valuable with their ability to selectively recalculate based on changes in specific signals. 
This means that a computed value or function can intelligently determine when to recalculate based on the signals it depends on. 

#### Caching

This approach involves the caching of results obtained from resource-intensive function calls. 
Instead of recalculating the output each time a function is invoked, the cached result is returned when the same inputs are encountered again. 

By remembering the last computed value and the conditions (or signal values) under which it was computed, the system can avoid unnecessary recalculations.

#### Dependency tracking

An essential aspect of effective memoization in a reactive system is accurate dependency tracking.  
This involves keeping track of the signals upon which a computed value relies. 
This way, the system can identify when the cached value needs to be invalidated and recomputed. 

This fine-grained dependency tracking ensures that the system's reactivity remains reliable and accurate.

### Store

Stores are important as they provide a structured and centralized approach to state management. 

> A store in reactive programming is essentially a container for state. 

Reviewer: Signals update "atomically", it is not possible to listen to only "part" of a signal. Stores, however, allow subscriptions to a specific part of the store, and will only trigger if that part of the store is updated. Therefore, stores are quite common when dealing with complex state.

Aside:
Internally, this is implemented using a tree of nested signals.

Aside:
When setting the store, it is important to only update the part of the store that has changed, as stores do not check whether data has actually changed before triggering. If you need to set a store without knowing exactly what changed, use the `reconcile` helper to diff the data.

> Stores are important as they provide a structured and centralized approach to state management. 
> 
> A store in reactive programming is essentially a container for state. 
> Unlike dealing with individual signals, stores offer a more organized way to manage state, especially in complex applications.

Author: ```suggestion
In reactive programming, stores play a crucial role by acting as specialized containers for managing information. 
They simplify the task of handling intricate scenarios within an app by centralizing data in an organized manner.

In contrast to signals, which update all at once without the option for selective listening, stores provide the advantage of targeted subscriptions.
With the ability to choose following specific parts of the state, updates will only be provided when those particular sections change.
This level of granularity makes stores important when managing complex state needs in applications.
```

Author: @mosheduminer let me know what you think about this change

## Importance of reactivity

1. **Automatic UI updates**: Reactivity ensures that as data changes, the user interface is automatically updated to reflect those changes.

2. **Simplified code**: By removing the need for manual DOM updates or UI refreshes, reactivity reduces the need for manual interventions. 
This synchronizes the state and the UI, leading to more straightforward and maintainable code.

3. **Consistent user experience**: With the UI always being in sync with the underlying data, users get a consistent and expected response to their actions.

4. **Scalability**: As applications grow, managing state and UI can become complex. Reactivity abstracts this complexity, ensuring that even large-scale applications remain manageable.

## Reactive principles

When delving into the world of reactivity, you'll encounter a few foundational principles:

### Signals

Signals serve as foundational elements in reactive systems, playing a pivotal role in data management and system responsiveness.

A signal can be visualized as a container for data. 
It not only holds the data, but also provides mechanisms for accessing and modifying its value, a **getter **and a **setter**. 

> This dual capability makes signals both readable and mutable, making them integral to the dynamic nature of reactive systems.

Reviewer 2: Maybe "readable and writable" is better phrasing 🤔 A signal is kinda like a variable, if a variable first points to a number, and then it points to another number, something changed but it feels weird to call that a "mutation". "Mutation" I think is generally used when talking about data structures, a signal is some kind of data structure but it's very similar to a variable, so it just feels a bit strange to me to use "mutable" in this context, but I guess it's perfectly fine also.

Author: No, you're right. Writeable is a better term than mutable

#### Getter

> The getter is the mechanism responsible for accessing the current value of the signal. 

Reviewer 2: Maybe just calling it a "function" rather than a "mechanism", or at least mentioning that it's a function somewhere, would be easier to understand. Like if I don't know what we are talking about already after reading this paragraph I don't know what to expect, "mechanism" sounds kinda complicated and convoluted.

When various components or parts of an application need to "read" or retrieve the data stored in a signal, they use the getter. 

> This provides data that is consistent and reliably accessible throughout the system.

Reviewer 2: This sentence seems superfluous, like reading it just adds confusion to me, now I'm thinking what other alternatives there could be that would give me data inconsistently or unreliably.

#### Setter

> On the other hand, the setter is the tool that permits the changing of the signal's value. 

Reviewer 2: Again "tool" like "mechanism" confuse me here, I think it's fine to call them this, but if one doesn't also say that they are functions then I'm totally in the dark here. Like the understanding that a "function" is a thing is probably a concept that any reader would have, because if they don't then there's nothing that we can say that could make them understand Solid anyway.

Any change to the value within a signal, made possible by the setter, can instigate reactive updates in other connected parts of the system. 
This ensures that the system remains responsive and up-to-date with the latest data changes.

### Effects

Effects and other listeners (such as memos) can be thought of as automated responders.
They are closely tied to signals and are set into motion whenever there's a change in the signals they observe. 
Their primary role is to ensure that any modifications in signals are appropriately reflected throughout the system, maintaining its reactivity.

The nature of effects revolves around two functionalities:

> - **Observation**: At the heart of every effect is its ability to continuously monitor or "observe" signals. 

Reviewer 2: "continuously monitor" makes me think of polling or something, which has nothing to do with how effects and memos work, but for an introduction to reactivity I guess that's doesn't matter.

This keeps effects primed to detect changes in the signals they are tied to.

> - **Activation**: Upon recognizing a change in an observed signal, the effect will promptly execute its predefined routine. 

Reviewer 2: "routine" I think is one of those words that means different things to different people 🤔 maybe just saying that it's refreshed or "called" could be easier to understand.

> This can involve tasks like updating the UI, triggering a process, or initiating external functions. 

Reviewer 2: "initiating" a function feels weird, a function can basically only be "called", not using that verb is confusing to me.

Through this activation, effects ensure that the system's reactivity is both immediate and relevant.

## Building a reactive system

> Reactive systems are designed around the principle of data-driven reactivity. This is when signals and their dependents, effects, maintain a relationship. This relationship ensures a seamless propagation of data changes throughout the system, guaranteeing its responsiveness and real-time updates.

Reviewer 2: Maybe worth saying "effects and memos", or the reader might think that memos are not dependents of signals.

### 1. Initialization

The start of a reactive system begins with the setup of signals. During this phase, signals are initialized with their respective starting values. These initialized signals serve as the primary data sources that the rest of the system will rely on.

### 2. Subscription

Once signals are established, the next step involves the effects indicating their dependencies. 
In this phase, effects "subscribe" to specific signals, marking their interest and establishing a link. 
This subscription keeps the effects aware of any changes to the signals they are tied to.

### 3. Notification

During the notification phase, any alteration in a signal's value sets off a chain reaction. 

> The altered signal proactively notifies all its subscribed effects about the modification. 

Reviewer 2: "altered" sounds a bit strange, maybe "updated" or "changed" is more natural, people don't usually say that data structures or variables are "altered".

Upon receiving this notification, these effects then update or activate as necessary to keep the system's state current and synchronized with the latest data changes.

## Synchronous vs. asynchronous

Reactivity in systems can manifest primarily in two ways: synchronous and asynchronous. 
The choice between these often depend on the requirements of the system and the nature of the tasks involved.

### Synchronous reactivity
Synchronous reactivity operates in a direct and linear fashion. 
When a signal changes, its corresponding effects are immediately updated in a sequential manner. 
This form of reactivity provides a direct correlation between a signal's change and the system's response, guaranteeing that the system remains consistently aligned with the most recent data.

### Asynchronous reactivity

> Asynchronous reactivity introduces a layer of flexibility and adaptability. 

Reviewer 2: A "layer of flexibility and adaptability" sounds strange to me, like I'm not sure what it's saying. Asynchronous/laziness is 100% about performance basically, so it's probably simpler to just say that "it can be faster in some cases" or something.

Unlike it synchronous counterpart, the updates to the effects are not instantaneous. 
Instead, they may be delayed or could wait for the completion of an external event or task.

This approach is useful in scenarios where an effect depends on multiple signals and when updating one signal before another may result a data inconsistency. 
In such a scenario, [`batch`](/reference/batch) can be used to delay triggering effects so the effect runs after both signals have been updated.

## Efficient data management

### Memo

Memo, short for memoization, is a technique that is used in reactive systems to enhance performance. 

In the context of reactivity, memos become valuable with their ability to selectively recalculate based on changes in specific signals. 
This means that a computed value or function can intelligently determine when to recalculate based on the signals it depends on. 

> 

Reviewer 2: It may be worth noting to that a memo is baaasically an effect that returns a signal to its return value. Like if one says that they are 95% the same thing then the whole system is easier to understand for the reader, imo.

#### Caching

This approach involves the caching of results obtained from resource-intensive function calls. 
Instead of recalculating the output each time a function is invoked, the cached result is returned when the same inputs are encountered again. 

By remembering the last computed value and the conditions (or signal values) under which it was computed, the system can avoid unnecessary recalculations.

#### Dependency tracking

An essential aspect of effective memoization in a reactive system is accurate dependency tracking.  
This involves keeping track of the signals upon which a computed value relies. 
This way, the system can identify when the cached value needs to be invalidated and recomputed. 

This fine-grained dependency tracking ensures that the system's reactivity remains reliable and accurate.

> 

Reviewer 2: Dependency tracking being under Memo feels weird, because it works the same both for effects and memos.

### Store

Stores are important as they provide a structured and centralized approach to state management. 

A store in reactive programming is essentially a container for state. 
Unlike dealing with individual signals, stores offer a more organized way to manage state, especially in complex applications.

#### State management

At their core, stores serve as containers for state. 

> They facilitate centralized management of state by offering methods to update, reset, or manipulate it. 

Reviewer 2: I don't think stores offer any method for resetting anything, like resetting implies that the original value is stored somewhere, but stores don't keep track of that kind of thing, so they can't provide a reset method.

> This not only improves the predictability of state transitions but simplifies the process of interacting with and modifying state.

Reviewer 2: Stores don't have anything to do about predictability, like by saying that I'm thinking that signals are less predictable or something. It's signals all the way down at the end of the day, the only thing that stores provide that signals don't give the user is convenience really.

When the state within a store changes, it triggers a cascade of reactive updates throughout the system. 
Components or functions that are subscribed to a store are promptly notified of these changes, where they are able to then respond accordingly. 
This keeps the user interface and other system components synchronized with the underlying data.

#### Scalability

As applications grow in complexity, managing individual signals can become challenging. 

> Stores provide a scalable solution by allowing related state to be grouped together. 

Reviewer 2: Exactly, they just make things simpler to handle.

## Tracking

At its core, tracking is about registering subscriptions. 

> When a signal changes, all subscribed effects are re-run. 

Reviewer 2: effects and memos.

This enables the automatic propagation of changes throughout an application's components.

Some key points to understand about tracking:

- Tracking subscriptions occur when a signal, or store value, is accessed within an effect. 
In other words, when you read the value of a signal inside an effect, that effect becomes subscribed to that signal.

- Whenever an effect is created and accesses a signal, it establishes a subscription to that signal. 
This means that the effect will be re-executed whenever the signal changes.

- Tracking is scoped to effects. 

> If you access a signal outside the context of an effect, no tracking or subscription will occur. 

Reviewer 2: That's not true, memos can subscribe to signals too.

Also it may be worth explicitly saying that fundamentally there's just something that can be changed (the signal), and something that can be updated (an effect or a memo). I think it would make the system way simpler to understand because one can understand immediately that if there's not something that can be changed then the app is basically static, and if there's nothing that can be updated then the app is static just the same. Like these "tools" are here to be able to write non-static stuff, with convenience and performance.

This is because there is no effect subscribed to that signal.

- The tracking concept extends to reactive props and store values, as well. 

> When you access a reactive value within an effect, it triggers tracking and subscription just like within signals.

Reviewer 2: This is maybe worth mentioning as is, but it's important to understand that only the "signal" exists that can be depended upon, "store values", "reactive pros" etc. don't really exist on their own, like it 100% boils down to whether a signal is called inside an effect/memo or not. How that signal is called is entirely irrelevant.

Within tracking, there are some key behaviours and tradeoffs that have to be considered:

> 1. All reactivity is tracked rom function calls directly, or hidden in the getter and triggered by the property's access. 

Reviewer 2: It's unclear what this is saying 🤔 I think it would be worth to show to different ways to call a signal inside an effect or memo, and then explicitly saying that it doesn't matter how the signal is called, it only matters where signals are called.

## Importance of reactivity

1. **Automatic UI updates**: Reactivity ensures that as data changes, the user interface is automatically updated to reflect those changes.

2. **Simplified code**: By removing the need for manual DOM updates or UI refreshes, reactivity reduces the need for manual interventions. 
This synchronizes the state and the UI, leading to more straightforward and maintainable code.

3. **Consistent user experience**: With the UI always being in sync with the underlying data, users get a consistent and expected response to their actions.

4. **Scalability**: As applications grow, managing state and UI can become complex. Reactivity abstracts this complexity, ensuring that even large-scale applications remain manageable.

[View on Eraser![](https://app.eraser.io/workspace/maDvFw5OryuPJOwSLyK9/preview?elements=wYq3tQ751U2WLyAHqYVMoA&type=embed)](https://app.eraser.io/workspace/maDvFw5OryuPJOwSLyK9?elements=wYq3tQ751U2WLyAHqYVMoA)

## Reactive principles

> When delving into the world of reactivity, you'll encounter a few foundational principles:

Reviewer 3: ```suggestion
When delving into the world of reactivity, you'll encounter a few foundational principles.
```

### Signals

> Signals serve as foundational elements in reactive systems, playing a pivotal role in data management and system responsiveness.

Reviewer 3: ```suggestion
Signals serve as core elements in reactive systems, playing a pivotal role in data management and system responsiveness.
```

Reviewer 3: You just used "foundational" in the previous sentence.

A signal can be visualized as a container for data. 

> It not only holds the data, but also provides mechanisms for accessing and modifying its value.

Reviewer 3: > its value

I'm not going to correct it, but "data" is technically plural. However, in modern English, it has also become singular.

- **Activation**: Upon recognizing a change in an observed signal, the effect will promptly be called. 
This can involve tasks like updating the UI, triggering a process, or calling external functions. 
Through this activation, effects ensure that the system's reactivity is both immediate and relevant.

[View on Eraser![](https://app.eraser.io/workspace/maDvFw5OryuPJOwSLyK9/preview?elements=XLztQwvdHs3ijzf9u2ZgPg&type=embed)](https://app.eraser.io/workspace/maDvFw5OryuPJOwSLyK9?elements=XLztQwvdHs3ijzf9u2ZgPg)

#### Dependency tracking

An essential part of managing effects in a reactive system is dependency tracking.  
This process involves keeping track of the specific signals that an effect relies on. 
By doing this, the system can determine when the effect needs to be re-triggered or updated based on changes in its dependent signals. 

This tracking is important when maintaining a fine-grained reactive system.
This level of precision in tracking dependencies helps in prevent unnecessary re-executions, contributing to a more reliable reactive system.

## Building a reactive system

Reactive systems are designed around the principle of data-driven reactivity. 

> This is when signals, effects, and memos, maintain a relationship. 

Reviewer 3: ```suggestion
This is when signals, effects, and memos maintain a relationship. 
```

### 1. Initialization

The start of a reactive system begins with the setup of signals. 
During this phase, signals are initialized with their respective starting values. 
These initialized signals serve as the primary data sources that the rest of the system will rely on.

### 2. Subscription

Once signals are established, the next step involves the effects indicating their dependencies. 
In this phase, effects "subscribe" to specific signals, marking their interest and establishing a link. 
This subscription keeps the effects aware of any changes to the signals they are tied to.

### 3. Notification

During the notification phase, any update in a signal's value sets off a chain reaction. 
The altered signal proactively notifies all its subscribed effects about the modification. 
Upon receiving this notification, these effects then update or activate as necessary to keep the system's state current and synchronized with the latest data changes.

## Synchronous vs. asynchronous

Reactivity in systems can manifest primarily in two ways: synchronous and asynchronous. 

> The choice between these often depend on the requirements of the system and the nature of the tasks involved.

Reviewer 3: ```suggestion
The choice between these often depends on the requirements of the system and the nature of the tasks involved.
```

### Synchronous reactivity
Synchronous reactivity operates in a direct and linear fashion. 
When a signal changes, its corresponding effects are immediately updated in a sequential manner. 
This form of reactivity provides a direct correlation between a signal's change and the system's response, guaranteeing that the system remains consistently aligned with the most recent data.

### Asynchronous reactivity
Asynchronous reactivity can improve performance in some cases. 
Unlike its synchronous counterpart, updates to effects are not instantaneous. 

> Instead, they may be delayed or could wait for the completion of an external event or task.

Reviewer 3: ```suggestion
Instead, they may be delayed, or they could wait for the completion of an external event or task.
```

This approach is useful in scenarios where an effect depends on multiple signals and when updating one signal before another may result a data inconsistency. 
In such a scenario, [`batch`](/reference/batch) can be used to delay triggering effects so the effect runs after both signals have been updated.

## Efficient data management

### Memo

Memo, short for memoization, is a technique that is used in reactive systems to improve performance.
In these systems, memos and effects share some similarities, with the primary difference being that memos *return* a signal as their return value.

> Furthermore, like effects, memos use dependency tracking  to make sure that recalculations and updates are triggered only when changes occur within their dependent signals.

Reviewer 3: ```suggestion
Furthermore, like effects, memos use dependency tracking to make sure that recalculations and updates are triggered only when changes occur within their dependent signals.
```

The value of a memo lies in its ability to selectively reclaculate; this happens only when there are changes in the dependent signals.
This process involves **caching** the results from resource-intensive function calls.

> Chaching stores the results of previous calculations, allowing the system to return a cached result rather than reclaculating when the same inputs are encountered.

Reviewer 3: ```suggestion
Caching stores the results of previous calculations, allowing the system to return a cached result rather than recalculating when the same inputs are encountered.
```

### Store

Stores are important as they provide a structured and centralized approach to state management. 

A store in reactive programming is essentially a container for state. 
Unlike dealing with individual signals, stores offer a more organized way to manage state, especially in complex applications.

#### State management

At their core, stores serve as containers for state, facilitating centralized state management by providing methods for updates and manipulations.
This makes changing state easier by simplifying interactions and modifications.

When the state within a store changes, it triggers a cascade of reactive updates throughout the system. 
Components or functions that are subscribed to a store are promptly notified of these changes, where they are able to then respond accordingly. 
This keeps the user interface and other system components synchronized with the underlying data.

#### Scalability

As applications grow in complexity, managing individual signals can become challenging. 
Stores provide a simple solution that allows related states to be grouped together. 

## Tracking

> At its core, tracking si fundamentally about managing subscriptions in a reactive system.

Reviewer 3: ```suggestion
At its core, tracking is fundamentally about managing subscriptions in a reactive system.
```

Signals serve as foundational elements in reactive systems, playing a pivotal role in data management and system responsiveness.

A signal can be visualized as a container for data. 
It not only holds the data, but also provides mechanisms for accessing and modifying its value.

Signals contain a dual capability that makes them both readable and writeable:
* **Getter**: 
The function responsible for accessing the current value of the signal.
When various components or parts of an application need to "read" or retrieve the data stored in a signal, they use a getter.
* **Setter**: 
The tool that permits the changing of the signal's value.
Any change to the value within a signal, made possible by the setter function, can create reactive updates in other connected parts of the system.
This ensures that the system remains responsive and up-to-date with the latest data changes.

> These aspects make them important to the dynamic nature of reactive systems.

Reviewer 2: I'm not sure what this sentence is saying.

Their primary role is to ensure that any modifications in signals are appropriately reflected throughout the system, maintaining its reactivity.

The nature of effects revolves around two functionalities:

- **Observation**: At the heart of every effect is its ability to "observe" or watch signals. 
This keeps effects primed to detect changes in the signals they are tied to.
- **Activation**: Upon recognizing a change in an observed signal, the effect will promptly be called. 
This can involve tasks like updating the UI, triggering a process, or calling external functions. 
Through this activation, effects ensure that the system's reactivity is both immediate and relevant.

#### Dependency tracking

An essential part of managing effects in a reactive system is dependency tracking.  
This process involves keeping track of the specific signals that an effect relies on. 
By doing this, the system can determine when the effect needs to be re-triggered or updated based on changes in its dependent signals. 

This tracking is important when maintaining a fine-grained reactive system.

> This level of precision in tracking dependencies helps in prevent unnecessary recalculations and executions, contributing to a more reliable reactive system.

Reviewer 2: what's the difference between recalculations and reexecutions?

This form of reactivity provides a direct correlation between a signal's change and the system's response, guaranteeing that the system remains consistently aligned with the most recent data.

### Asynchronous reactivity

> Asynchronous reactivity can be improve performance in some cases. 

Reviewer 2: there's an extra "be" here

#### State management

At their core, stores serve as containers for state, facilitating centralized state management by providing methods for updates and manipulations.

> This improves state transitions by streamlining the process of interacting with and altering state.

Reviewer 2: This sentence seems pretty confusing to me, I'm not sure what its saying. And "to alter" is not a thing that people say about state I think, like I've never heard it use elsewhere that I can remember anyway.

#### Scalability

As applications grow in complexity, managing individual signals can become challenging. 
Stores provide a simple solution that allows related states to be grouped together. 

## Tracking

At its core, tracking si fundamentally about managing subscriptions in a reactive system.

> When a signal undergoes a change, all the effects *and* memos subscribed to it are automatically re-run; this ensures the propogation of changes throughout the application's components.

Reviewer 2: maybe adding a "re-run if needed" could be worth it.

### Key concepts

* Subscriptions are initaited when a signal, or a store value, are accessed.
This interaction leads to the accessor subscribing to the signal, establishing a dynamic connection.

> Every interaction between the signal and its accessor reinforces this subscription, guaranteeing that the associated effects are re-executed when the signal changes.

Reviewer 2: subscriptions can't be reinforced, either they exist or they don't.

* Both memos *and* effects have the ability to subscribe to signals.

> However, interactions with a signal outside of these won't initiate tracking, due to the absence of a subscriber, thus limiting the scope of reactivity.

Reviewer 2: it may not be clear that memos and effects are the subscribers here 🤔

* The reactive system is built on the interplay between elements that can change (signals) and those that can be updated (effects and memos).
The principles of tracking are not limited to these elements, however, but also extend to reactive props and store values.
* Entities like store values and reactive props are inherently depdendent on signals.

> A "signal" stands as the only entity that operates autonomously and is reliably dependable.

Reviewer 2: Confusing sentence, like what other entities are not reliably dependable? Maybe saying "objects" rather than "entities" would be simpler. And "operate autonomously" seems a way more complicated way of saying "they don't depend on anything", which I'm not if it's even worth saying 🤔 maybe one can say that signals are the leaves in the reactivity graph.

Using Solid:
```tsx
function Counter() {
  const [count, setCount] = createSignal(0);
  const increment = () => setCount(count() + 1); 

  createEffect(() => {
    console.log("My effect says " + count());
  })
    /* With the `createEffect`, the `count()` (the signal) is now tracked as a subscription of this effect. This code will run *only* when `count()` changes. */

  console.log(count()); // In this case,`count()` is *not* in a trackintg scope. This means it won't re-run when `count()` changes.

  return (
    <button type="button" onClick={increment}>
      {count()} 

>       {/* JSX is a tracking scope because it uses effects behind the scenes. Because of this, the code will register `count()` as a subscription when it is called. */}

Reviewer 2: maybe saying that it "wraps expressions in effects" rather than a generic "uses effects" could be understandable, but I guess it should be fine either way.
